src/expiring_lru.rs

```diff
@@ -64,39 +42,83 @@ where
     type Target = V;
 
     fn deref(&self) -> &Self::Target {
-        match &self.0 {
-            EntryRefInner::RefMut(inner) => &inner.value().inner,
-            EntryRefInner::RefMulti(inner) => &inner.value().inner,
-        }
+        &self.0.value().inner
     }
 }
 
-async fn expire_entries<K, V>(map: ExpiringLru<K, V>, reap_interval: Duration, expiration: Duration)
-where
-    K: Eq + Hash,
+async fn decay_task<K, V>(
+    map: Arc<DashMap<K, Entry<V>>>,
+    expiration: Duration,
+    mut rx: UnboundedReceiver<TimerUpdate<K, V>>,
+) where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Send + Sync + 'static,
 {
-    let mut interval = interval(reap_interval);
+    let mut queue = DelayQueue::new();
 
     loop {
-        interval.tick().await;
-
-        let right_now = Instant::now();
-        map.inner
-            .retain(|_, entry| entry.last_used + expiration > right_now);
-
-        debug!("Done reaping timed out HTTP ratelimiters");
+        tokio::select! {
+            Some(key) = queue.next(), if !queue.is_empty() => {
+                // An item expired in the queue, remove it from the map
+                debug!("Removing expired entry from ratelimiter decay queue");
+                map.remove(key.get_ref());
+            }
+            Some(msg) = rx.recv() => {
+                match msg {
+                    TimerUpdate::Add { key, value } => {
+                        debug!("Adding entry to ratelimiter decay queue");
+                        let decay_key = queue.insert(key.clone(), expiration);
+                        let entry = Entry {
+                            inner: value,
+                            decay_key,
+                        };
+                        map.insert(key, entry);
+                    },
+                    TimerUpdate::Refresh { key } => {
+                        debug!("Refreshing entry in ratelimiter decay queue");
+                        // This will panic if the key is not present, therefore
+                        // we check that in the calling end
+                        queue.reset(&key, expiration);
+                    },
+                    TimerUpdate::Remove { key } => {
+                        debug!("Removing entry in ratelimiter decay queue");
+                        queue.try_remove(&key);
+                    }
```

Comment:
Branch its associated method are never used, could we just remove them?

Reply:
We could, but I thought I'd keep it for the sake of the completeness of the implementation. Will remove it